fiber/adapter.go

```diff
@@ -85,7 +85,7 @@ func (f *FiberLambda) adaptor(w http.ResponseWriter, r *http.Request) {
 	req.SetHost(r.Host)
 	for key, val := range r.Header {
 		for _, v := range val {
-			req.Header.Add(key, v)
+			req.Header.Set(key, v)
```

Comment:
Why do you use `set`? It will override header with the same key.

Reply:
Add cannot pass header['Content-Type' or 'User-Agent'.. ] to fiber app

Reply:
fasthttp/~~/header.go


``` go
// Add adds the given 'key: value' header.
//
// Multiple headers with the same key may be added with this function.
// Use Set for setting a single header for the given key.
func (h *ResponseHeader) Add(key, value string) {
	k := getHeaderKeyBytes(&h.bufKV, key, h.disableNormalizing)
	h.h = appendArg(h.h, b2s(k), value, argsHasValue)
}

```

``` go
// Use Add for setting multiple header values under the same key.
func (h *RequestHeader) Set(key, value string) {
	initHeaderKV(&h.bufKV, key, value, h.disableNormalizing)
	h.SetCanonical(h.bufKV.key, h.bufKV.value)
}
.....



// SetCanonical sets the given 'key: value' header assuming that
// key is in canonical form.
func (h *RequestHeader) SetCanonical(key, value []byte) {
	switch string(key) {
	case HeaderHost:
		h.SetHostBytes(value)
	case HeaderContentType:
		h.SetContentTypeBytes(value)
	case HeaderUserAgent:
		h.SetUserAgentBytes(value)
	case HeaderCookie:
		h.collectCookies()
		h.cookies = parseRequestCookies(h.cookies, value)
	case HeaderContentLength:
		if contentLength, err := parseContentLength(value); err == nil {
			h.contentLength = contentLength
			h.contentLengthBytes = append(h.contentLengthBytes[:0], value...)
		}
	case HeaderConnection:
		if bytes.Equal(strClose, value) {
			h.SetConnectionClose()
		} else {
			h.ResetConnectionClose()
			h.h = setArgBytes(h.h, key, value, argsHasValue)
		}
	case HeaderTransferEncoding:
		// Transfer-Encoding is managed automatically.
	default:
		h.h = setArgBytes(h.h, key, value, argsHasValue)
	}
}

```

It's quite differnet

Reply:
Ops. It's quite different Add and Set.

`Add` method use [appendArg](https://github.com/valyala/fasthttp/blob/4637395dd7bcf16ad812036bb465fbc34d2697a0/args.go#L433)

`Set` method use [setArgBytes](https://github.com/valyala/fasthttp/blob/4637395dd7bcf16ad812036bb465fbc34d2697a0/args.go#L408) => [appendArg]

Add is soooooooo simple to add header. But, Set method has more kitty functions.

Reply:
> `Add` cannot pass header['Content-Type' or 'User-Agent'.. ] to fiber app

Emm... do you have debug info? Like print all request headers.

Reply:
> It will override header with the same key.

fiber.Ctx cannot access Header Key[User-Agent, Content-Type]. You can check my test codes.

Sorry, I don't know why it'll override header. Would you plz explain that?

Reply:
> Emm... do you have debug info? Like print all request headers. 
Oh, Sure. Wait minutes.

Reply:
Http.Header is an alias for `map[string][]string`. It means a header key can have multiple values. That's why we use `Add` but not `Set`.

Reply:
### when use 'Add' method
![image](https://user-images.githubusercontent.com/10516961/109821802-b127e180-7c79-11eb-95de-17ad80eb731e.png)

output: 
``` text
 ~/Dropbox/drakejin/aws-lambda-go-api-proxy/fiber   master ● ?  ginkgo -v                                                                                                                 ✔  6037  23:39:07
Running Suite: Fiber Suite
==========================
Random Seed: 1614782348
Will run 1 of 1 specs

FiberLambda tests Simple ping request 
  Proxies the event correctly
  /Users/drakejin/Dropbox/drakejin/aws-lambda-go-api-proxy/fiber/fiberlambda_test.go:17
2021/03/03 23:39:10 Starting test
2021/03/03 23:39:10 ------------------ kiyon's request start -------------------
2021/03/03 23:39:10 Just print all header: When use **Add** method
2021/03/03 23:39:10 GET /ping HTTP/1.1
Content-Length: 0
User-Agent: Agent Gopher/1.16
Content-Type: application/json; charset=utf-8


2021/03/03 23:39:10 but, can not get 'Content-Type' or 'User-Agent' ...
2021/03/03 23:39:10 Content-Type :  []
2021/03/03 23:39:10 User-Agent :  []
2021/03/03 23:39:10 ------------------ kiyon's request end -------------------
2021/03/03 23:39:10 Handler!!
2021/03/03 23:39:10 ------------------ kiyon's request start -------------------
2021/03/03 23:39:10 Just print all header: When use **Add** method
2021/03/03 23:39:10 GET /ping HTTP/1.1
Content-Length: 0
User-Agent: Agent Gopher/1.16
Content-Type: application/json; charset=utf-8
X-Golambdaproxy-Apigw-Stagevars: null
X-Golambdaproxy-Apigw-Context: {"accountId":"","resourceId":"","stage":"","domainName":"","domainPrefix":"","requestId":"","protocol":"","identity":{"cognitoIdentityPoolId":"","accountId":"","cognitoIdentityId":"","caller":"","apiKey":"","apiKeyId":"","accessKey":"","sourceIp":"","cognitoAuthenticationType":"","cognitoAuthenticationProvider":"","userArn":"","userAgent":"","user":""},"resourcePath":"","authorizer":null,"httpMethod":"","requestTime":"","requestTimeEpoch":0,"apiId":""}


2021/03/03 23:39:10 but, can not get 'Content-Type' or 'User-Agent' ...
2021/03/03 23:39:10 Content-Type :  []
2021/03/03 23:39:10 User-Agent :  []
2021/03/03 23:39:10 ------------------ kiyon's request end -------------------
•2021/03/03 23:39:10 Handler!!

Ran 1 of 1 Specs in 0.005 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 1.678671298s
Test Suite Passed
```

Reply:
### when i use 'Set' method
![image](https://user-images.githubusercontent.com/10516961/109822211-17146900-7c7a-11eb-82da-36446d159034.png)

``` text
 ~/Dropbox/drakejin/aws-lambda-go-api-proxy/fiber   master ● ?  ginkgo -v                                                                                                                 ✔  6038  23:40:59
Running Suite: Fiber Suite
==========================
Random Seed: 1614782509
Will run 1 of 1 specs

FiberLambda tests Simple ping request 
  Proxies the event correctly
  /Users/drakejin/Dropbox/drakejin/aws-lambda-go-api-proxy/fiber/fiberlambda_test.go:17
2021/03/03 23:41:51 Starting test
2021/03/03 23:41:51 ------------------ kiyon's request start -------------------
2021/03/03 23:41:51 Just print all header: When use **Set** method
2021/03/03 23:41:51 GET /ping HTTP/1.1
User-Agent: Agent Gopher/1.16
Content-Type: application/json; charset=utf-8
Content-Length: 0


2021/03/03 23:41:51 but, can not get 'Content-Type' or 'User-Agent' ...
2021/03/03 23:41:51 Content-Type :  [97 112 112 108 105 99 97 116 105 111 110 47 106 115 111 110 59 32 99 104 97 114 115 101 116 61 117 116 102 45 56] => application/json; charset=utf-8
2021/03/03 23:41:51 User-Agent :  [65 103 101 110 116 32 71 111 112 104 101 114 47 49 46 49 54] => Agent Gopher/1.16
2021/03/03 23:41:51 ------------------ kiyon's request end -------------------
2021/03/03 23:41:51 Handler!!
•2021/03/03 23:41:51 ------------------ kiyon's request start -------------------
2021/03/03 23:41:51 Just print all header: When use **Set** method
2021/03/03 23:41:51 GET /ping HTTP/1.1
User-Agent: Agent Gopher/1.16
Content-Type: application/json; charset=utf-8
Content-Length: 0
X-Golambdaproxy-Apigw-Stagevars: null
X-Golambdaproxy-Apigw-Context: {"accountId":"","resourceId":"","stage":"","domainName":"","domainPrefix":"","requestId":"","protocol":"","identity":{"cognitoIdentityPoolId":"","accountId":"","cognitoIdentityId":"","caller":"","apiKey":"","apiKeyId":"","accessKey":"","sourceIp":"","cognitoAuthenticationType":"","cognitoAuthenticationProvider":"","userArn":"","userAgent":"","user":""},"resourcePath":"","authorizer":null,"httpMethod":"","requestTime":"","requestTimeEpoch":0,"apiId":""}


2021/03/03 23:41:51 but, can not get 'Content-Type' or 'User-Agent' ...
2021/03/03 23:41:51 Content-Type :  [97 112 112 108 105 99 97 116 105 111 110 47 106 115 111 110 59 32 99 104 97 
114 115 101 116 61 117 116 102 45 56] => application/json; charset=utf-8
2021/03/03 23:41:51 User-Agent :  [65 103 101 110 116 32 71 111 112 104 101 114 47 49 46 49 54] => Agent Gopher/1.16
2021/03/03 23:41:51 ------------------ kiyon's request end -------------------
2021/03/03 23:41:51 Handler!!
Ran 1 of 1 Specs in 0.001 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 1.636651574s
Test Suite Passed
```

Reply:
test code is PR's

Reply:
> test code is PR's

Thanks for this info! That's really weird. I'll investigate it tomorrow!

Reply:
> Http.Header is an alias for map[string][]string

Good point. I knew that. But, the fiber app is not using a `http.Header`. It use `fasthttp.Header`. And fasthttp.Header isn't designed like hash map.

Reply:
> > Http.Header is an alias for map[string][]string
> 
> 
> 
> But, the fiber app is not using a `http.Header`. It use `fasthttp.Header`. And fasthttp.Header isn't designed like hash map.

When using `Add`, it'll behavior like the hash map. 

For example 
```go
req.Header.Add("K1", "v1")
req.Header.Add("K1", "v2")
```
Header will be 
```
K1: v1
K1: v2
```

And http.Header also have `Add` and `Set` if I remember correctly.

Reply:
Yes, you're right

``` go
**type RequestHeader struct {
	noCopy noCopy //nolint:unused,structcheck

	disableNormalizing bool
	noHTTP11           bool
	connectionClose    bool

	// These two fields have been moved close to other bool fields
	// for reducing RequestHeader object size.
	cookiesCollected bool

	contentLength      int
	contentLengthBytes []byte

	method      []byte
	requestURI  []byte
	proto       []byte
	host        []byte
	contentType []byte
	userAgent   []byte

	h     []argsKV  // <-------- this member variable do you said, but it doesn't help when access 'Content-Type' LOL. What the hack. I think fasthttp need to change. I cannot understand why they designed it.
	bufKV argsKV

	cookies []argsKV

	// stores an immutable copy of headers as they were received from the
	// wire.
	rawHeaders []byte
}
**
```

Reply:
You are right. I think the best way is adding these special headers
```go
req.SetHost(r.Host)
req.Header.SetUserAgent(r.UserAgent())
req.Header.SetContentType(r.Header.Get(fiber.HeaderContentType))
```

Reply:
With out change method? 
``` go
   --- req.Header.Add(key, v)
   +++ req.Header.Set(key, v)
```

Reply:
Yes

Reply:
Okay, I'll commit following your guide.

Reply:
Sorry, I think more about this. It's much better to use `Set` than `Add`.
There's no reason to use `Add` method in this situation.

You have to see https://github.com/awslabs/aws-lambda-go-api-proxy/pull/89#discussion_r586473703 this outputs. `Set` 

> Http.Header is an alias for map[string][]string. It means a header key can have multiple values. That's why we use Add but not Set.

I think you already know working on `Set` method. `Set` method already include all features about `Add`. So I cannot understand why doing like `SetCanonical` method things. It's already designed by `fasthttp`.

[SetCanonical](https://github.com/valyala/fasthttp/blob/master/header.go#L984-L1016) This method is included at `Set`

Please check again @kiyonlin If you think still my opinion is not a good solution, I'll following your guide.
respect you.

Reply:
I think you haven't understand `Add` yet, I wish this test case can make it clear:
```go
It("Proxies the event correctly", func() {
	app := fiber.New()
	app.Get("/ping", func(c *fiber.Ctx) error {
		Expect(c.Get(fiber.HeaderUserAgent)).To(Equal("fiber"))
		Expect(c.Get(fiber.HeaderContentType)).To(Equal(fiber.MIMEApplicationJSON))
		c.Context().Request.Header.VisitAll(func(key, value []byte) {
			if string(key) == "K1" {
				_, _ = c.Write(value)
			}
		})
		return nil
	})

	adapter := fiberadaptor.New(app)

	req := events.APIGatewayProxyRequest{
		Path:       "/ping",
		HTTPMethod: "GET",
		MultiValueHeaders: map[string][]string{
			fiber.HeaderUserAgent:   {"fiber"},
			fiber.HeaderContentType: {fiber.MIMEApplicationJSON},
			"k1":                    {"v1", "v2"},
		},
	}

	resp, err := adapter.Proxy(req)

	Expect(err).To(BeNil())
	Expect(resp.StatusCode).To(Equal(200))
	Expect(resp.Body).To(Equal("v1v2"))
})
```

You can try to use `Set` and see what will happen.